### PR TITLE
query(fuse): limit push down respect orders

### LINKED ISSUE #4807: bug: If use SELECT LIMIT query a table that the data is not batch insert the result of query will not stable
### Search before asking

- [X] I had searched in the [issues](https://github.com/datafuselabs/databend/issues) and found no similar issues.

### Version

The newest main commit.

Server version: 8.0.26-v0.2.1-alpha-74587a0-simd(rust-1.61.0-nightly-2022-04-12T06:13:57.952366803Z) 0

### What's Wrong?

The limit result for a table is not stable, the result usually changed.

But if I create a table and the data is selected from numbers() the limit result always true.

### How to Reproduce?

Start server with:

```
$ make run
```

#### Prepare table with table engine FUSE:

```SQL
create table t1(id int);
insert into t1 values(1);
insert into t1 values(2);
insert into t1 values(3);
insert into t1 values(4);
mysql> select * from t1;
+------+
| id   |
+------+
|    4 |
|    3 |
|    2 |
|    1 |
+------+
4 rows in set (0.06 sec)
Read 4 rows, 16 B in 0.006 sec., 658.47 rows/sec., 2.63 KB/sec.

```

try to execute limit query:

**The result of `order by id asc limit 1,2` should be [2,3].**

```SQL
root@mysqldb 12:50:16 [(none)]> select * from t1 order by id asc;
+------+
| id   |
+------+
|    1 |
|    2 |
|    3 |
|    4 |
+------+
4 rows in set (0.06 sec)
Read 4 rows, 16 B in 0.007 sec., 596.2 rows/sec., 2.38 KB/sec.

root@mysqldb 12:50:28 [(none)]> select * from t1 order by id asc limit 1,2;
+------+
| id   |
+------+
|    2 |
|    4 |
+------+
2 rows in set (0.06 sec)
Read 3 rows, 12 B in 0.007 sec., 421.53 rows/sec., 1.69 KB/sec.

root@mysqldb 12:50:32 [(none)]> select * from t1 order by id asc limit 1,2;
+------+
| id   |
+------+
|    2 |
|    4 |
+------+
2 rows in set (0.06 sec)
Read 3 rows, 12 B in 0.007 sec., 441.06 rows/sec., 1.76 KB/sec.

root@mysqldb 12:50:34 [(none)]> select * from t1 order by id asc limit 1,2;
+------+
| id   |
+------+
|    3 |
|    4 |
+------+
2 rows in set (0.06 sec)
Read 3 rows, 12 B in 0.007 sec., 449.21 rows/sec., 1.8 KB/sec.

mysql> select * from t1 order by id asc limit 2 offset 1;
+------+
| id   |
+------+
|    3 |
|    4 |
+------+
2 rows in set (0.06 sec)
Read 3 rows, 12 B in 0.007 sec., 411.61 rows/sec., 1.65 KB/sec.

mysql> select * from t1 order by id asc limit 2;
+------+
| id   |
+------+
|    2 |
|    4 |
+------+
2 rows in set (0.06 sec)
Read 2 rows, 8 B in 0.006 sec., 332.02 rows/sec., 1.33 KB/sec.

--After optimize table t1, the result is not correct.
mysql> select * from t1 order by id asc limit 2;
+------+
| id   |
+------+
|    3 |
|    4 |
+------+
2 rows in set (0.06 sec)
Read 2 rows, 8 B in 0.006 sec., 319.84 rows/sec., 1.28 KB/sec.


```

The result of `order by id desc limit 1,2` should be [3,2]

```SQL
root@mysqldb 12:50:38 [(none)]> select * from t1 order by id desc limit 1,2;
+------+
| id   |
+------+
|    3 |
|    1 |
+------+
2 rows in set (0.06 sec)
Read 3 rows, 12 B in 0.007 sec., 439.25 rows/sec., 1.76 KB/sec.

root@mysqldb 12:52:42 [(none)]> select * from t1 order by id desc limit 1,2;
+------+
| id   |
+------+
|    3 |
|    2 |
+------+
2 rows in set (0.03 sec)
Read 3 rows, 12 B in 0.007 sec., 430.44 rows/sec., 1.72 KB/sec.

```

**But if use this case, the limit query result is always correct**

```SQL
root@mysqldb 12:56:39 [(none)]> create table t(number int);
Query OK, 0 rows affected (0.05 sec)
Read 0 rows, 0 B in 0.010 sec., 0 rows/sec., 0 B/sec.

root@mysqldb 12:56:46 [(none)]> insert into t select number from numbers(5) order by number limit 1,4;
Query OK, 0 rows affected (0.05 sec)
Read 5 rows, 40 B in 0.013 sec., 391.77 rows/sec., 3.13 KB/sec.

root@mysqldb 12:56:48 [(none)]> select * from t;
+--------+
| number |
+--------+
|      1 |
|      2 |
|      3 |
|      4 |
+--------+

root@mysqldb 12:58:08 [(none)]> select * from t order by number desc limit 1,2;
+--------+
| number |
+--------+
|      3 |
|      2 |
+--------+
2 rows in set (0.06 sec)
Read 4 rows, 16 B in 0.006 sec., 624.07 rows/sec., 2.5 KB/sec.

root@mysqldb 12:58:30 [(none)]> select * from t order by number asc limit 1,2;
+--------+
| number |
+--------+
|      2 |
|      3 |
+--------+
2 rows in set (0.05 sec)
Read 4 rows, 16 B in 0.006 sec., 631.21 rows/sec., 2.52 KB/sec.

```

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by sundy-li**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

Summary about this PR

## Changelog
 
- Bug Fix
 

## Related Issues

Fixes #4807

### CODE REVIEW

**Review by dantengsky — APPROVED** [review]
LGTM

**Comment by TCeason** [comment]
That’s cool

File: query/src/storages/fuse/operations/read_partitions.rs [hunk 1]
```diff
@@ -75,10 +75,13 @@ impl FuseTable {
         blocks_metas: &[BlockMeta],
         push_downs: Option<Extras>,
     ) -> (Statistics, Partitions) {
-        let limit = push_downs
-            .as_ref()
-            .and_then(|p| p.limit)
-            .unwrap_or(usize::MAX);
+        let mut limit = usize::MAX;
```
Comment (Xuanwo) [inline query/src/storages/fuse/operations/read_partitions.rs:78]: How about:

```rust
let limit = push_downs
            .as_ref()
            .filter(|p| p.order_by.is_empty())
            .and_then(|p| p.limit)
            .unwrap_or(usize::MAX);
```

Comment (sundy-li) [reply query/src/storages/fuse/operations/read_partitions.rs:78]: > How about:
> 
> ```rust
> let limit = push_downs
>             .as_ref()
>             .filter(|p| p.order_by.is_empty())
>             .and_then(|p| p.limit)
>             .unwrap_or(usize::MAX);
> ```

That's great, I did not come up with this idea.

**Review by Xuanwo — APPROVED** [review]

**Pull request merged**